### PR TITLE
Add DNS compress pointer mutation option, and some minor changes.

### PR DESCRIPTION
**Pull request by Pentiumluyu**

### CODE REVIEW

File: src/chinadns.c [hunk 1]
```diff
@@ -622,11 +696,9 @@ static void queue_add(id_addr_t id_addr) {
...
 static id_addr_t *queue_lookup(uint16_t id) {
   int i;
-  // TODO assign new id instead of using id from clients
-  for (i = 0; i < ID_ADDR_QUEUE_LEN; i++) {
-    if (id_addr_queue[i].id == id)
-      return id_addr_queue + i;
-  }
+  i = (uint8_t)(id & 0xff);
+  if (id_addr_queue[i].id == id)
+  	return id_addr_queue + i;
```
Comment (clowwindy) [inline src/chinadns.c]: `id` is picked by random rather than increment in glibc. If we use sharding here like this, we have a high collision rate here: 50% collision rate in 16 concurrent queries(compared to 256 concurrent queries before the change). This is not enough if we ChinaDNS is running on a LAN that have several users. We should use precise comparison instead. My suggestion is to undo the change here.

File: src/chinadns.c [hunk 2]
```diff
@@ -669,8 +741,15 @@ static int should_filter_query(ns_msg msg, struct in_addr dns_addr) {
     dns_is_foreign = !dns_is_chn;
   }
   rrmax = ns_msg_count(msg, ns_s_an);
-  if (rrmax == 0)
+  if (rrmax == 0){
+    if (compression){
```
Comment (clowwindy) [inline src/chinadns.c]: space required between `)` and `{`

File: src/chinadns.c [hunk 3]
```diff
@@ -683,10 +762,11 @@ static int should_filter_query(ns_msg msg, struct in_addr dns_addr) {
...
       if (verbose)
         printf("%s, ", inet_ntoa(*(struct in_addr *)rd));
-      r = bsearch(rd, ip_list.ips, ip_list.entries, sizeof(struct in_addr),
+      if (!compression)
+        r = bsearch(rd, ip_list.ips, ip_list.entries, sizeof(struct in_addr),
                   cmp_in_addr);
-      if (r)
-        return 1;
+        if (r)
+          return 1;
```
Comment (clowwindy) [inline src/chinadns.c]: wrong indentation or missing `{ }`?

**Comment by clowwindy** [comment]
Thanks. I've added some notes, please check them.

File: src/chinadns.c (hunk 3 above)
Comment (Pentiumluyu) [inline src/chinadns.c]: typo...using tabs here...

File: src/chinadns.c [hunk 4]
```diff
@@ -683,10 +764,12 @@ static int should_filter_query(ns_msg msg, struct in_addr dns_addr) {
     if (type == ns_t_a) {
       if (verbose)
         printf("%s, ", inet_ntoa(*(struct in_addr *)rd));
-      r = bsearch(rd, ip_list.ips, ip_list.entries, sizeof(struct in_addr),
+      if (!compression)
+        r = bsearch(rd, ip_list.ips, ip_list.entries, sizeof(struct in_addr),
```
Comment (clowwindy) [inline src/chinadns.c]: I mean here:

```
      if (!compression)
        r = bsearch(rd, ip_list.ips, ip_list.entries, sizeof(struct in_addr),
                  cmp_in_addr);
        if (r) {
          return 1;
        }
```

This is actually:

```
      if (!compression)
        r = bsearch(rd, ip_list.ips, ip_list.entries, sizeof(struct in_addr),
                  cmp_in_addr);

      if (r) {
        return 1;
      }
```

Comment (Pentiumluyu) [inline src/chinadns.c]: Yeah, you're right...

**Comment by clowwindy** [comment]
Merged, thanks!

**Comment by clowwindy** [comment]
Released 1.3.0
